### ISSUE TITLE
Integrity constraint violation with Project Config

### ISSUE THREAD

**Issue by engram-design** [angell-co/Spoon#90]
Had the following issue after editing blocks locally, and pushing to production.

```
Applying changes from project.yaml ... error: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'matrixBlockTypeId' cannot be null
The SQL being executed was: INSERT INTO `spoon_blocktypes` (`fieldId`, `matrixBlockTypeId`, `groupName`, `context`, `uid`, `fieldLayoutId`, `dateCreated`, `dateUpdated`) VALUES (127, NULL, 'Content', 'global', '577d299d-31b9-4451-b793-05bb8b88726e', 238, '2020-02-27 22:21:10', '2020-02-27 22:21:10')
```

This also left 10 out of 13 block types 'reset' so I had to set them up through Spoon again.

**Comment by qrazi** [comment]
Running into this too, Craft CMS 3.4.9 and Spoon 3.5.2. The matrixBlockType in question is new, defined in `project.yaml` by another developer. So probably for some reason Spoon tries to add a record that is not added by Craft yet (as I cannot find this record in my DB. But that could also be due to failed transactions?).

**Comment by qrazi** [comment]
Should probably do a check similar to: https://github.com/craftcms/cms/blob/develop/src/services/Matrix.php#L325

**Comment by joshangell** [comment]
@qrazi you hero, thanks for finding that! If you want to PR it that would be a big help.

**Comment by qrazi** [comment]
@joshangell I am still testing / investigating on my end, but I'll try and isolate this bug and do a PR (the issues on my end include more things than just this one... 😬 )

**Comment by joshangell** [comment]
@qrazi aaaah sorry about that, there are an increasing number of edge case bugs popping up that are frustratingly time consuming to fix!

**Comment by qrazi** [comment]
@joshangell No worries! I mean in the project I am working, not specifically this plugin 😅

**Comment by qrazi** [comment]
Ok, things are up running on my environment again, and right now I do not need the check similar to the Matrix-service. The other issues I kept running into when syncing the `project.yaml` were cause by duplicate MatrixBlockFields. 

I am still not a 100% why that ended up in Spoon trying to insert a record before the accompanying MatrixBlockField was inserted, but probably because when searching based on handle it would have given a result somewhere. The Craft CMS ecosystem uses both handles and uids. Not that long ago this used to be solely based on handles, so duplicating handles was actually a lot more difficult. We're now in this mixed code that on the one hand allows create duplicate handles, but on the other hand always expects a handle to be unique.

I am not sure if the cause is similar for @engram-design error? I will create a PR anyway, but I think it might solve a symptom, not a cause.

**Comment by engram-design** [comment]
So, I believe what happened at the time, was the Matrix blocks themselves didn't update, or something went wrong, and due to that, Spoon tried to do its thing without having the Matrix blocks properly sorted first.

It's all a bit vague, but it was pretty urgent to fix at the time, so had to revert things, and manually set things up from scratch.

**Comment by qrazi** [comment]
> So, I believe what happened at the time, was the Matrix blocks themselves didn't update, or something went wrong, and due to that, Spoon tried to do its thing without having the Matrix blocks properly sorted first.
> 
> It's all a bit vague, but it was pretty urgent to fix at the time, so had to revert things, and manually set things up from scratch.

Yes, we've been doing that a lot too. It's really easy to get these kind of issues in Craft CMS. Development discipline remains key. However, the PR would point developers a bit more in the right direction I think.

**Comment by davist11** [comment]
Was there any sort of resolution to this problem? I'm seeing the same thing happen that it's trying to sync the spoon config before the block types

**Comment by dennisfrank** [comment]
I am running into this too. @engram-design Have you checked PR #93?

**Comment by Collinvdv** [comment]
Same problem here..

**Comment by dennisfrank** [comment]
?

**Comment by engram-design** [comment]
@dennisfrank Haven't tried the PR - did it work for you?

**Comment by cherrykoda** [comment]
Just ran into this myself - any word?

**Comment by davist11** [comment]
We've started to switch some projects over to https://plugins.craftcms.com/matrixmate since there hasn't been much action here. Just in case that could be helpful to anyone.

**Comment by joshangell** [comment]
I can finally reproduce this! Will be aiming to fix asap.

**Comment by joshangell** [comment]
OK - seems like the PR #93 fixes this - have tested against my reproducible case an it fixes that, though I’m not going to rule out it not fixing other situations so please do re-open this if that is the case.
